sympy/geometry/util.py

```diff
@@ -684,23 +681,44 @@ def intersection(*entities):
     """
     from .entity import GeometryEntity
     from .point import Point
-
-    if len(entities) <= 1:
-        return []
-
-    # entities may be an immutable tuple
+    pairwise = kwargs.pop('union', False)
+
+    if pairwise == False:
+
+        if len(entities) <= 1:
+            return []
+
+        # entities may be an immutable tuple
+        entities = list(entities)
+        for i, e in enumerate(entities):
+            if not isinstance(e, GeometryEntity):
+                try:
+                    entities[i] = Point(e)
+                except NotImplementedError:
+                    raise ValueError('%s is not a GeometryEntity and cannot be made into Point' % str(e))
+
+        res = entities[0].intersection(entities[1])
+        for entity in entities[2:]:
+            newres = []
+            for x in res:
+                newres.extend(x.intersection(entity))
+            res = newres
+        return res
+
+    ans = []
     entities = list(entities)
     for i, e in enumerate(entities):
         if not isinstance(e, GeometryEntity):
             try:
                 entities[i] = Point(e)
             except NotImplementedError:
                 raise ValueError('%s is not a GeometryEntity and cannot be made into Point' % str(e))
-
-    res = entities[0].intersection(entities[1])
-    for entity in entities[2:]:
-        newres = []
-        for x in res:
-            newres.extend(x.intersection(entity))
-        res = newres
-    return res
+    for j in range(0, len(entities)):
+        for k in range(j + 1, len(entities)):
+            if len(intersection(entities[j], entities[k])) is not 0:
```

Comment:
Maybe we should just have a new function, too, and make the keyword on intersection private (i.e. `_pairwise`) to discourage its use: 
```
def pairwise_intersections(*objs):
    """Return all unique pairwise intersections that occur between the given objects.
    return intersection(*objs, pairwise=True)
```

And here is a test:

```
c, l, s, t = Circle((0, 0), 1), Line((0, 0), slope=oo), Segment((0, 0), (0, 1)), Triangle((0, 1), (2, 0), (-2, 0))
assert pairwise_intersections(c, l, s, t) == [
    Point2D(-1, 0), Point2D(-4/5, 3/5), Point2D(0, 1), Point2D(4/5, 3/5), Point2D(1, 0), s]
```